### ISSUE TITLE
Binding modes/binding submaps

### ISSUE THREAD

**Issue by YaLTeR** [YaLTeR/niri#846]
The ability to configure multiple binding modes with separate binds.

An example use-case is a "launcher" mode, where e.g. you press Super+D, then instead of the normal binds, you have binds to run applications, like F for firefox, N for nautilus, T for terminal, etc. So you press Super+D+F to run firefox.

There's a WIP implementation for binding modes: https://github.com/YaLTeR/niri/pull/532. The author's GitHub account had disappeared, but I remember about the PR and will get to it at some point.

**Comment by mastoca** [comment]
xmonad's approach for subkeys is interesting which I used in many ways. for example I use it to bring up the gridSelect module
```        ^++^ subKeys
          "Grid Select (CTRL-g followed by a key)"
          [ ("M1-g g", addName "grid select some bookmarks" $ spawnSelected' myBookmarkGrid),
```

**Comment by bbb651** [comment]
In my opinion instead of binding modes we should allow bindings to have multiple keys (like in @mastoca's example). This is more intuitive and matches other existing software (e.g. vim/neovim, vscode, zed, emacs, ghostty). They are both identical in functionally since they just represent a state machine, but multikey syntax lets bindings be local which is good for managing and sharing configs, and hands off the task of building the state machine to niri.

I've started working on this, I've done the parsing and the syntax works really nicely by (ab)using arguments:
```kdl
Mod+G G cooldown=150 {
    // ...
}
```

<details>
    <summary>Code</summary>

```rs
        let keys =
            iter::once(node.node_name.parse::<Key>().map_err(|e| {
                DecodeError::conversion(&node.node_name, e.wrap_err("invalid keybind"))
            }))
            .chain(node.arguments.iter().map(|val| {
                match &*val.literal {
                    knuffel::ast::Literal::String(s) => Ok(s),
                    _ => Err(DecodeError::unexpected(
                        &val.literal,
                        "argument type",
                        "expected argument to be a string",
                    )),
                }?
                .parse::<Key>()
                .map_err(|e| {
                    DecodeError::conversion(&node.node_name, e.wrap_err("invalid keybind"))
                })
            }))
            .collect::<Result<Vec<Key>, DecodeError<_>>>()?;

```

</details>

This requires reworking bindings to a state machine, I've currently going with a naive approach of storing the index of the "active key" for each bind incrementing it when a key matches (going from longest to shortest sequence to allow bindings that end with the entire sequence of keys of another bind) and resetting all the indices once a bind is triggered, we can switch to a trie if this turns out to be a performance issue.
I currently need to refactor everything that touches `mods_with_binds` to use the state machine.

I'm not sure what should happen when an unrelated key is pressed while a multikey binding is ongoing, should it:
- Reset binding state, send the key?
- Reset binding state, supress the key?
- Keep binding state, send the key?
- Keep binding state, suppress the key? Probably not, easy to get stuck in a binding

We also need to think about how this interacts with release binds.

**Comment by YaLTeR** [comment]
Hmm. What I don't like about this approach is that it's infectious in a way: you put one multi-key bind somewhere in your config (and forget about it), and suddenly pressing the first key just stops doing anything, or starts working with a big latency (the timeout). On the other hand, binding modes are much more explicit about this. You can also be explicit about the unrelated keys, since probably in different cases it makes sense to do different things about them.

**Comment by titaniumtraveler** [comment]
I have been thinking about topic submaps a lot this week, and actually think there might be a nice way to marry the concepts of submaps and multikey binds.

Basically every keybind is associated with a submap (set as a property defaulting to `default`) and we then keep track of what submap we are in.
(Potentially that could be a *stack* of submaps, which would allow for different kind of overlay maps)

Multikey binds could then just be implemented by creating a submap for that key.

To make that not to inconvenient/unintuitive to the user bindings could be nested to automatically create and enter that submap:

```kdl
Mod+D {
    F { spawn "firefox";   }
    N { spawn "nautilus";  }
    T { spawn "alacritty"; }
}
```

This would then automatically create a submap named `Mod+D` and add the binds `F`, `N` and `T` to it and a bind `Mod+D` to enter the submap.

I would be happy to implement this.
I depend on submaps a lot in my workflow, so having them in `niri` would be nice 🙂

**Comment by luleyleo** [comment]
So basically anonymous submaps? I quite like the idea, as most of my Hyprland submaps could be replaced with that. I think that would provide the best balance between the convenience of multi-key-bindings and the flexibility of named submaps.

**Comment by titaniumtraveler** [comment]
> So basically anonymous submaps? I quite like the idea, as most of my Hyprland submaps could be replaced with that. I think that would provide the best balance between the convenience of multi-key-bindings and the flexibility of named submaps.

I mean they *are* nameable. (Just by executing an action like `enter_submap "Mod+D";`)
But you don't have to explicitly invent a name for them.

Also The nesting would only be *one* way to put a keybind into a submap.
I think something
```kdl
Mod+S submap=custom {

}
```
should be possible as well.

But yeah, I really didn't like the way hyprland handled submaps.
They always felt like a bunch of overhead if what you really wanted where just a chorded keymap.

**Comment by gleachkr** [comment]
FWIW, I've been using https://github.com/MaxVerevkin/wlr-which-key, which works great for binding submaps. Maybe there's a case to be made for keeping the core niri binding mechanisms simple, and dispatching to another tool when you want submaps.

**Comment by bphenriques** [comment]
> FWIW, I've been using https://github.com/MaxVerevkin/wlr-which-key, which works great for binding submaps. Maybe there's a case to be made for keeping the core niri binding mechanisms simple, and dispatching to another tool when you want submaps.

I am using it and fits like a glove for keyboard centric flows. However, I miss some features that I need to share with the devs: dynamic labels (e.g., current sink when selecting next one) or show/hide dynamically (e.g., hide a audio sink if it is not enabled atm). 

This goes a bit beyond of what might be desirable in Niri as it can get complex.

**Comment by bphenriques** [comment]
Update on my feedback: it would be great for `wlrctl` as it would allow moving the mouse with the keyboard without a overlay as `wlr-which-key` does. Look forward for it.

https://github.com/moverest/wl-kbptr?tab=readme-ov-file#hyprland

**Comment by bbb651** [comment]
@bphenriques I actually have a WIP [mouseless](https://youtu.be/J0rwQVNQkHM) clone I'm working on, I was waiting for wlr-virtual-pointer because uinput is very junky for absolute position, it was very quick to port over (still not very usable due to performance, especially with text):

https://github.com/user-attachments/assets/2ea6d5f9-078e-4c28-81f0-121fce6ae7f5

But off topic for this issue

**Comment by bphenriques** [comment]
> [@bphenriques](https://github.com/bphenriques) I actually have a WIP [mouseless](https://youtu.be/J0rwQVNQkHM) clone I'm working on, I was waiting for wlr-virtual-pointer because uinput is very junky for absolute position, it was very quick to port over (still not very usable due to performance, especially with text):
> clip-1741865813.mp4
> 
> But off topic for this issue

Nice! I am currently using ` wl-kbptr`. I will monitor your project. Yeah, slighty offtopic but related

**Comment by idlip** [comment]
Just dropping the message since I thought of some feature when keychords/submaps are implemented.

The feature is inspired based on `universal-argument` from emacs (invoking prefix as C-u)
The idea is, with universal argument we can spawn any window in current workspace and as float with no effect on window rules.

It will be very useful when having many windows and columns, so irrespective of the window rules it would be nice if we can spawn any window in current screen/workspace in floating mode just for quick glance.

This might be relevant to scratch window...

Thank you for considering this!

**Comment by idlip** [comment]
@bbb651 @bphenriques 
For moving mouse, and other inputs, there is `dotool` which works in wayland as well.

https://git.sr.ht/~geb/dotool

Hope that helps

**Comment by YaLTeR** [comment]
> Just dropping the message since I thought of some feature when keychords/submaps are implemented.

This is more of a #915, orthogonal to binding modes.

**Comment by YaLTeR** [comment]
> FWIW, I've been using https://github.com/MaxVerevkin/wlr-which-key, which works great for binding submaps. Maybe there's a case to be made for keeping the core niri binding mechanisms simple, and dispatching to another tool when you want submaps.

This has come across my attention again. Seems like a fairly good impl, and comes with a GUI too.

Do we really need submaps in niri given that it exists?

> However, I miss some features that I need to share with the devs: dynamic labels (e.g., current sink when selecting next one) or show/hide dynamically (e.g., hide a audio sink if it is not enabled atm).

niri impl would not have these features either. Perhaps it's better to use wlr-which-key with scripts or contribute there?

I would very much like not increasing the amount of code and complexity that I need to deal with in niri itself, if perfectly viable app-side solutions exist.

**Comment by bphenriques** [comment]
> Do we really need submaps in niri given that it exists?

Yes, I use `wlr-which-key` but I wish GUI was hidden in some cases (e.g., when moving my mouse around) or when I set multiple keybindings. Sometimes the GUI gets a bit overwhelming. I can elaborate more but essentially, there are "shortcut modes" that I wish it was hidden and does not fit `wlr-which-key`. 

My setup is [here](https://github.com/bphenriques/dotfiles/blob/17d14fadb3d8e989eae7abced100367d88c86e4c/config/home-manager/desktop-environment/wlr-which-key.nix#L31). 

Examples

![Image](https://github.com/user-attachments/assets/f47b4c69-225f-47af-a3ad-d0253c9f8c2d)

![Image](https://github.com/user-attachments/assets/4aab9400-1fc7-4d1e-8d15-f74acd21742f)

**Comment by YaLTeR** [comment]
Sounds like this is much more easily fixed in wlr-which-key than adding a whole submaps system to niri?

**Comment by bphenriques** [comment]
> Sounds like this is much more easily fixed in wlr-which-key than adding a whole submaps system to niri?

Maybe, up for the author to decide if it makes sense tbh. Can't judge that. Something to suggest and let him decide.

**Comment by YaLTeR** [comment]
Could patch for own use in the meantime

**Comment by bphenriques** [comment]
> Could patch for own use in the meantime

Yeah. Need to find time to learn Rust and propose changes but life gets in the way. For now I am _okay_.

**Comment by tmccombs** [comment]
wlr-which-key is nice, and definitely works for some use cases. However, it doesn't fully replace a binding mode. 


For example, with a binding mode it is possible to switch into a resize mode where using directional keys (or hjkl) adjusts the size of the current window. Or a move mode, where you can use the keyboard to move a floating window.

That said, it probably would be possible for such a thing to be implemented in a separate application, that captured keyboard input, as long as you are ok with not being able to send regular key events to other apps while in that mode. 

It would certainly be more convenient to have it built in to niri though.

**Comment by YaLTeR** [comment]
> For example, with a binding mode it is possible to switch into a resize mode where using directional keys (or hjkl) adjusts the size of the current window. Or a move mode, where you can use the keyboard to move a floating window.

This can be implemented in a client like wlr-which-key too. By just not hiding it on the first key press, but keeping it up until some cancel bind like Escape.

**Comment by FreeFull** [comment]
> > For example, with a binding mode it is possible to switch into a resize mode where using directional keys (or hjkl) adjusts the size of the current window. Or a move mode, where you can use the keyboard to move a floating window.
> 
> This can be implemented in a client like wlr-which-key too. By just not hiding it on the first key press, but keeping it up until some cancel bind like Escape.

Except for special cases such as the screenshot GUI. Currently you can use the window move/resize bindings, but I find those rather inconvenient, it'd be nice to be able to just press arrow down to move the area down, and so on. With a separate screenshot binding mode, it'd also be possible to set how much the area gets moved/resized by by each binding

**Comment by tmccombs** [comment]
Doh! Somehow I totally missed the keep_open option

**Comment by gro-david** [comment]
This would actually be really useful. For example I have Ctrl+Space bound to my launcher, but any time I try to do some gaming trying to sprint and jump at the same time just triggers my launcher.

**Comment by tmccombs** [comment]
That could potentially be addressed by having a way to temporarily turn off all keybinds except for one that turns them back on. Or maybe re-use the allow-when-locked setting? Although, I'm not sure how that would work with exiting the mode.

**Comment by YaLTeR** [comment]
Well that's called keyboard shortcuts inhibit but it has to be initiated by the client via the protocol at the moment

**Comment by tmccombs** [comment]
Oh, right, I knew there was something like that, but couldn't find it. So maybe allow the `toggle-keyboard-shortcuts-inhibit` action to enable keyboard inhibiting as well as disabling it? Or have an action that allows turning on keyboard inhibiting for a client that hasn't requested it itself?

**Comment by YaLTeR** [comment]
Maybe

**Comment by psa-anddev** [comment]
Just a comment to show interest in this issue to be addressed. I've been checking Niri and I love the concept and the idea. I see it as the evolution of Hyprland. However, I think the addition of modes would be great. I like to set my keybindings to be set using pnemonics using the Super key for every shortcut that has to do with the compositor. So, for instance, I use Super+c to enter a communication mode in which, I can use s to open Signal, e to open Element or m to open mutt. Without a system of modes, keybindings can become difficult to use and that can be seen in the default config with the keybinding Mod+Shift+Ctrl+h to move a window to the monitor in the left. I have an ErgoDox EZ and configuring keys to do be Mod+Shift+Ctrl can ease the problem, however, without a programmable keyboard, this kind of key bindings can cause problems in the hands. By experience, I can say that I move windows between monitors in a regular basis. 

I think this feature would make Niri even better. Yes, there are ways to get around the lack of this feature (at the moment, I've taken the suggestion of using wlr-which-key), it would be great to have all my keybindings in one place. 

Again, I really love the concept of Niri and I appreciate all the work you are doing.

**Comment by rickswe** [comment]
Take inspiration from Xmonad, they did a good job. It’s been years though since I used it. I think you can do this with Keyd but it would be more of an external dependency.

**Comment by FoolsTech** [comment]
I'm migrating to from Hyprland and submaps are a huge deal breaker to me. So huge thanx to @gleachkr for mentioning wlr-which-key. But while it does the job very good, it's a bit hacky. For example you can't get rid off the gui popup. I think it needs to run without a gui for the first 100-200ms and only then popup like it's done in neovim. But it seems to be not possible. So basically my point is I like how it's done in Neovim's Which-Key. And to counter @YaLTeR's reservations about such method: I never run into the issues you mention and I have an unhealthy amount of multikey binds in Neovim (as well as in Hyprland), because I can easily check what binds already exist natively and via user configs with one or two commands. So it would need in addition to the core functionality a conflict detection and a cli action to search for existing key binds I guess.

So how are the chances to merge a pull-request on this matter? Did you decide whether niri shall have submaps/multikey binds? @YaLTeR 

By the way wlr-which-key's `--initial-keys` option doesn't work with the version installed via cargo for some reason. I had to compile it on my own and to fix some errors along the way.

**Comment by YaLTeR** [comment]
> For example you can't get rid off the gui popup. I think it needs to run without a gui for the first 100-200ms and only then popup like it's done in neovim. But it seems to be not possible.

It's possible by changing the relevant logic in wlr-which-key.

> So how are the chances to merge a pull-request on this matter? Did you decide whether niri shall have submaps/multikey binds? @YaLTeR 

I haven't decided yet.

Again my main reservation is that this is a lot of complexity in the compositor (which brings down all applications if it crashes for example, or locks you out of all applications if it breaks, unlike a client like wlr-which-key, not to mention I get to maintain the complexity) while the same, or very close to the same, functionality can be done in a client (and in fact is already done in wlr-which-key, baring various minor tweaks).

**Comment by FoolsTech** [comment]
> I haven't decided yet.
> 
> Again my main reservation is that this is a lot of complexity in the compositor (which brings down all applications if it crashes for example, or locks you out of all applications if it breaks, unlike a client like wlr-which-key, not to mention I get to maintain the complexity) while the same, or very close to the same, functionality can be done in a client (and in fact is already done in wlr-which-key, baring various minor tweaks).

ok, good to know, thanx.

**Comment by TWB0109** [comment]
Hey, I know this adds a lot of complexity, I just want to say that I'm interested in this happening as I love niri but that's the only thing I'm missing. Specially for small keyboard users it's a bit complicated to use a compositor without submaps. 

I guess it could be fixed by using something like keyd or keymonad though.

**Comment by FoolsTech** [comment]
> Hey, I know this adds a lot of complexity, I just want to say that I'm interested in this happening as I love niri but that's the only thing I'm missing. Specially for small keyboard users it's a bit complicated to use a compositor without submaps.
> 
> I guess it could be fixed by using something like keyd or keymonad though.

@TWB0109 
wlr-which-key does the job quite well. I have set up 300 keybinds using it. works flawlessly and partially better than submaps in hyprland. The only negative thing is, that it doesn't block further submap calls while one is already listening for key press, so that I sometimes end up with two submap overlays on the screen. But I think it is easy to fix, just need to get familiar with the code base (which is not huge though with 1.5k LoC).

**Comment by TIAcode** [comment]
wlr-which-key doesn't really do what's possible in other WMs. Like I used to have special submap for remote desktop sessions - I'd still be able to change the window or workspace on the host with special keyboard shortcuts, but most of the keys would get passed through to the remote desktop. Even something like "resize mode" where you'd enter the mode and then use the keys multiple times to adjust the window aren't possible (though that probably wouldn't be that hard to implement)

**Comment by bartlibert** [comment]
@TIAcode Just to get you on the way, I have a resize mode in wlr-which-key with this config (keep_open is the key!)
``` 
- key: "r"
    desc: Resize
    submenu:
      - key: "h"
        desc: Shrink width
        cmd: niri msg action set-column-width "-5%"
        keep_open: true
      - key: "l"
        desc: Expand width
        cmd: niri msg action set-column-width "+5%"
        keep_open: true
      - key: "j"
        desc: Shrink height
        cmd: niri msg action set-window-height "-5%"
        keep_open: true
      - key: "k"
        desc: Expand height
        cmd: niri msg action set-window-height "+5%"
        keep_open: true

```

**Comment by wlritchi** [comment]
There's one important way that multi-key / sequence bindings can't replace modal bindings: they don't work for running a series of keybinds in the same mode. For example, I want to have a mode for moving columns around (e.g. `Mod+M` to enter movement mode, then `up up up left` to move the column three workspaces up and one to the left, then `Esc` to exit movement mode). Sequence bindings could only get me as far as `Mod+M up Mod+M up Mod+M up Mod+M left`, which isn't what I'm looking for. In short, sequence bindings implicitly return to the default mode after they match.

As @bartlibert mentions, `wlr-which-key` and other submap-based keybind systems can solve this with something like a `keep_open` flag on submaps.

On the flipside, if you want to emulate sequence bindings using modal bindings, you can do that as long as the binding language supports something like `{ <action>; binding-mode default; }`.

**Comment by Cu3PO42** [comment]
I also use submaps extensively and am having trouble getting into Niri without them. While much of my workflow could probably be mapped to a client such as wlr-which-key, one thing I don't see how to solve is how I can emulate submaps with a client while still allowing focus on other windows.

It is my understanding that wlr-which-key (and any tool like it) needs to create a surface that is focused for it to receive and process keyboard events. I don't think I could keep using my applications as normal while that is the case, could I? Maybe the accessibility interface could be used over D-Bus to solve this?

Alternatively, a protocol extension could allow any client the necessary access, but that seems problematic from a security standpoint. Then again, is the same not true for the D-Bus interface?

**Comment by bertbesser** [comment]
Reading this discussion, and checking the docs for wlr-which-key, I also came to the conclusion that wlr-which-key can cover for parts of what I want, but still
- requires more (and maybe redundant keystrokes),
- takes focus, which can be in the way,
- and is a bit more complex to configure.

So
1. thanks for pointing me to wlr-which-key, which I might check out,
2. I'd still be pleased to see submaps built into niri (with some of the typical features like automatically leaving a submap after a timeout with no further strokes and such).

And, very importantly:
Thanks for niri! I'm "spiking" using it, and it's a joy so far.

Cheers

**Comment by bertbesser** [comment]
Short feedback. Tried wlr-which-key, and for the time being it bridges the gap quite well for me. 👍🏼 So thanks again!

**Comment by milanglacier** [comment]
[> wlr-which-key doesn't really do what's possible in other WMs. Like I used to have special submap for remote desktop sessions - I'd still be able to change the window or workspace on the host with special keyboard shortcuts, but most of the keys would get passed through to the remote desktop. Even something like "resize mode" where you'd enter the mode and then use the keys multiple times to adjust the window aren't possible (though that probably wouldn't be that hard to implement)


I’d like to echo this. More precisely, `wlr-which-key` cannot implement a true passthrough mode—one that allows all keys defined in the main mode to be forwarded to a VM or remote desktop. This would make it possible to switch workspaces or change focus within the VM or remote session using the same keybindings as on the host machine, simply by activating the passthrough mode.

**Comment by YaLTeR** [comment]
> I’d like to echo this. More precisely, `wlr-which-key` cannot implement a true passthrough mode—one that allows all keys defined in the main mode to be forwarded to a VM or remote desktop. This would make it possible to switch workspaces or change focus within the VM or remote session using the same keybindings as on the host machine, simply by activating the passthrough mode.

The VM app should request Wayland shortcuts inhibiting, then this will already work just fine.

**Comment by tmccombs** [comment]
> The VM app should request Wayland shortcuts inhibiting, then this will already work just fine.

If it is a remote session inside a browser, is it even possible for the web app to request that?

**Comment by YaLTeR** [comment]
I am not familiar with available JavaScript APIs

**Comment by Cu3PO42** [comment]
There is no standard API to request shortcut inhibition in the browser, unfortunately. Recent Chromium uses the shortcut inhibition protocol if you are in fullscreen and call an experimental Chromium-only `keyboard.lock()` API, but it's not guaranteed behavior and rather an implementation detail.

**Comment by YaLTeR** [comment]
Well, this doesn't need submaps. This needs our toggle shortcut inhibit action to be able to enable it on windows that didn't request it, in addition to disable on windows that did request it

**Comment by milanglacier** [comment]
> Well, this doesn't need submaps. This needs our toggle shortcut inhibit action to be able to enable it on windows that didn't request it, in addition to disable on windows that did request it

Thanks for the prompt reply! I hadn’t noticed that the example keybinding already includes `Mod+Escape allow-inhibiting=false { toggle-keyboard-shortcuts-inhibits };`. I agree that if this toggle inhibition action can properly handle misbehaving applications that fail to implement the keyboard shortcut request correctly, then a submap for passthrough behavior isn’t necessary.

**Comment by clotodex** [comment]
I generally agree with @YaLTeR. The reason I am switching from hyprland is that it grew into a monster that keeps crashing and breaks compatibility constantly. (Plus was using hyprscrolling and a first-class-citizen approach here is really nice)
Anyways a few things I am breaking my brain about (in no particular order):
- **modularity** (like xwayland-sattelite): I like this philosophy and even if wlr-which-key does not cover all usecases, putting together an effort that solves the submap problem compositor independent is really really nice to have
- **native integration & speed**: [_EDIT: this is more important than I thought. Outsourcing this requires a root-level tool capturing input from evdev or uinput (e.g. remappers). These exist but need to be managed separately by design and are often conflicting. Wlr-which-key only achieves its functionally with UI (it is a window listening to keybinds) - so actually not having a UI is not possible. To me this shows that while the complexity of the state machine should/could be outsourced, the integration of third party tools is not easy at all._]
I personally use submaps mainly for compositor-internal things (resizing or the "movement-mode" mentioned). Outsourcing keypresses has a potential danger in my mind of not being snappy enough. This point might be mute though (did not experience problems yet with wlr-which-key)
- **compatibility**: given hyprland I worry. Having to do ANY action via `niri msg` requires every action to be accessible (this is true as far as i can tell though) and any change would break the wlr-which-key config. **EDIT:** also ipc usage in bars etc. Does it make sense for e.g. wlr-which-key to also build up an ipc or dbus system? (working on https://github.com/MalpenZibo/ashell/pull/328)
- **accessibility / multiple-input-devices**: a big point for me to use niri was also its support for touchscreens, and generally multiple ways of input. Outsourcing would (at least for now) separate different input methods from each other. So if niri brings first-class support to more touchpad, touchscreen, mouse interactions - they wont carry over. However, this is still early in niri as well can could be brought back if relevant? And would prevent a whole lot of complexity for now
- **compositor independency / coupling**: this to me is the biggest headache. I really like outsourcing things that dont have to be tied to a specific compositor - does not lock me in. But since I use submaps (personally) a LOT for interacting only with that specific compositor, it feels like an unnecessary redirection in use. My wlr-which-key config is not transferrable and needs to be kept in perfect sync with niri (see compatibility). Plus it adds the burden on the user to keep two different configs of keymaps maintained (which leads me to think if ALL keybinds could be outsourced or if that would make sense at all)

I hope these points help the discussion. I love niri for having a quite flawless out-of-the-box experience (with documentation), that makes switching to it easy. Still trying to judge though how many people (however strong we care) are actually submap users :D

**Comment by onion108** [comment]
Heavily uses submap for sequenced keybinds. I immediately stopped thinking about moving to niri as I found there is no such feature exists.

**Comment by jloic** [comment]
I use modal keybindings for tmux, vim ... the only missing in the nesting sequence is niri now. Also, with a minimal keyboard it's cool to not rely on 2/3 modifiers for each bindings.

**Comment by WizardUli** [comment]
I have a question for users of `wlr-which-key` on Niri:
When trying to create a focus "mode" using `keep_open: true` which stays on until I cancel it with Esc key, I encountered a problem when crossing monitor boundaries. When I start `wlr-which-key` with this config on my leftmost monitor
```
inhibit_compositor_keyboard_shortcuts: true

menu:
  - key: "h"
    desc: hhh
    cmd: niri msg action focus-monitor-left
    keep_open: true
  - key: "l"
    desc: lll
    cmd: niri msg action focus-monitor-right
    keep_open: true
```
and press "l" the focus keeps traveling through all my monitors to the rightmost one and `wlr-which-key` stops intercepting all keys including Esc so I have to kill it. Presumably it stops intercepting key releases as well that's why it keeps repeating the action.

How do you solve this?

**Comment by mageowl** [comment]
@WizardUli Please open a new issue for `wlr-which-key`. That is unrelated to this feature request.

**Comment by WizardUli** [comment]
> [@WizardUli](https://github.com/WizardUli) Please open a new issue for `wlr-which-key`. That is unrelated to this feature request.

It think it is probably a very nice example why handling keypress sequences belongs to WM and not a 3rd party app which breaks when the focus leaves the monitor where it resides.

Moreover even desktop-environments/shells like noctalia-shell leave keypress handlings to the VM (whether it is Niri or Hyperland or SwayWM).

**Comment by aahsnr-configs** [comment]
@WizardUli  I am the process of using xremap-niri package from the aur. I am mostly using as a replacement for submaps in Hyprland. Should I expect trouble with my bindings then? I am trying to move from Hyprland to Niri

**Comment by ImYuugen** [comment]
Hello, has there been any progress on this issue ? Like a WIP somewhere or anything ? I would be interested in trying to implement this (can not promise anything)

**Comment by Sempyos** [comment]
Uhh not really

**Comment by ImYuugen** [comment]
Then I wonder if sir maintainer would be okay with merging a potential implementation of this.

I have a use case that was not mentionned here (and I don't think wlr-which-key fixes ?)
On hyprland, you can use `pass` to send keystrokes to unfocused windows, this way you can have something similar to global keybinds.
In my case, I used submaps to send keystrokes to a program, for example an `obs` submap would send all keys to OBS.

I don't think this kind of behaviour is possible using a client (however I know that Niri doesn't support that "`pass`" feature yet, but it could be in preparation for the future) so that's at least one argument in favour of implementing it.

Let's talk about it ! I want to know if it's possible before starting to code

**Comment by bra0** [comment]
https://github.com/bra0/niri-mods

**Comment by akileshas** [comment]
hey folks, how are subscribed to this feature, an note to you.
in recent weeks i have been thinking and designing the thinks for having submap's in-built
in niri-wm, which could be essential for the users who are jumping from hyprland to niri.

if you could, please give out your suggestions to improve and push out the design and implementation further.
also, if you could please test out the things and give your input to fix any of the issues you have encountered.
here is the drafted PR - #4604 .
